pkg/controllers/flannelmigration/network_migrator.go

```diff
@@ -0,0 +1,191 @@
+	return &networkMigrator{
+		ctx:          ctx,
+		calicoClient: calicoClient,
+		k8sClientset: k8sClientset,
+		config:       config,
+	}
+}
+
+// Initialise network migrator.
+func (m *networkMigrator) Initialise() error {
+	// Set calico image
+	d := daemonset(m.config.CalicoDaemonsetName)
+	image, err := d.GetContainerImage(m.k8sClientset, namespaceKubeSystem, calicoNodeContainerName)
+	if err != nil {
+		return err
+	}
+	m.calicoImage = image
+
+	// Set calico CNI config file name
+	cniConf, err := d.GetContainerEnv(m.k8sClientset, namespaceKubeSystem, calicoCniContainerName, calicoCniConfigEnvName)
+	if err != nil {
+		return err
+	}
+	m.calicoCNIConfigName = cniConf
+
+	log.Infof("Network migrator initialised, container image %s, CNI config %s/%s.", m.calicoImage, m.config.CNIConfigDir, m.calicoCNIConfigName)
+	return nil
+}
+
+// Remove Flannel network device/routes on node.
+// Write a dummy calico cni config file in front of Flannel/Canal CNI config.
+// This will prevent Flannel CNI from running and make sure new pod created will not get networked
+// until Calico CNI been correctly installed.
+func (m *networkMigrator) removeFlannelNetworkAndInstallDummyCalicoCNI(node *v1.Node) error {
+	// Run a remove-flannel pod with specified nodeName, this will bypass kube-scheduler.
+	// https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodename
+
+	// Deleting a tunnel device will remove routes, APR and FDB entries related with the device.
+	// Deleting cni0 device to remove routes to local pods.
+	// It is possible tunnel device or cni0 has been deleted already.
+	dummyCNI := `{ "name": "dummy", "plugins": [{ "type": "flannel-migration-in-progress" }]}`
+
+	cmd := fmt.Sprintf("{ ip link show cni0; ip link show flannel.%d; } || exit 0 && { echo '%s' > /host/%s/%s ; ip link delete cni0 type bridge; ip link delete flannel.%d; } && exit 0 || exit 1",
+		m.config.FlannelVNI, dummyCNI, m.config.CNIConfigDir, m.calicoCNIConfigName, m.config.FlannelVNI)
+	pod := k8spod("remove-flannel")
+	podLog, err := pod.RunPodOnNodeTillComplete(m.k8sClientset, namespaceKubeSystem, m.calicoImage, node.Name, cmd, m.config.CNIConfigDir, true, true)
+	if podLog != "" {
+		log.Infof("remove-flannel pod logs: %s.", podLog)
+	}
+
+	// Wait twice as long as default sync period so kubelet has picked up dummy CNI config.
+	// TODO we probably need something better here.
+	time.Sleep(2 * defaultSyncConfigPeriod)
+	return err
+}
+
+// Drain node, remove Flannel and setup Calico network for a node.
+func (m *networkMigrator) setupCalicoNetworkForNode(node *v1.Node) error {
+	log.Infof("Setting node label to disable Flannel daemonset pod on %s.", node.Name)
+	// Set node label so no Flannel pod can be scheduled on this node.
+	// Flannel pod currently running on the node starts to be evicted as the side effect.
+	n := k8snode(node.Name)
+	err := n.addNodeLabels(m.k8sClientset, nodeNetworkUnknown)
```

Comment:
Should "nodeNetworkUnknown" really be called "nodeNetworkNone"? It seems it's meant to signify "Don't run anything on this node", right?

Reply:
Sounds good.